cljfmt/src/cljfmt/core.cljc

```diff
@@ -273,6 +274,21 @@
   ([form indents]
    (indent (unindent form) indents)))
 
+(defn- trailing-whitespace? [zloc]
```

Comment:
I think this function is a little overly complex. What we want is to match two circumstances:

1. Whitespace followed by a newline
2. Whitespace that's at the end of the file

That sound about right? So we could write (1) as:

```clojure
(defn- trailing-whitespace? [zloc]
  (let [r (zip/right zloc)]
    (and (whitespace? zloc) (line-break? r) (not (comment? r))))
```

As you likely discovered for yourself, `line-break?` matches both newlines and comments:

```clojure
(defn- line-break? [zloc]
  (or (zlinebreak? zloc) (comment? zloc)))
```

That was because I didn't think there'd be cause to treat them differently. However, in this case we do want to treat them differently, so we can just use the raw `zlinebreak?` function (which probably needs a better name, now).

```clojure
(defn- trailing-whitespace? [zloc]
  (and (whitespace? zloc) (zlinebreak? (zip/right zloc))))
```

That takes care of (1). Now we just need to figure out (2). It turns out that if we just use:

```clojure
(nil? (zip/right zloc))
```

Then we fall afoul of whitespace internal to forms, e.g. `( foo )` becomes `( foo)`. I see you noticed this as well, because you added in `(not (surrounding? zloc whitespace?))`.

However, I think there's a better way of handling this. We check `(nil? (zip/right zloc))` only when we know we're directly below the root form:

```clojure
(defn final? [zloc]
  (and (nil? (zip/right zloc)) (root? (zip/up zloc))))
```

How do we know what the root form is? Well, when we can no longer go `zip/up`:

```clojure
(defn root? [zloc]
  (nil? (zip/up zloc)))
```

So all together:

```clojure
(defn- trailing-whitespace? [zloc]
  (and (whitespace? zloc)
       (or (zlinebreak? (zip/right zloc)) (final? zloc))))
```

This also removes the need for `trimr` in `reformat-string`.